src/model/engines/engine_base.nit

```diff
@@ -53,5 +76,88 @@ class Engine
 	fun compile(submission: Submission): Bool is abstract
 
 	# Run `test` for `submission`
-	fun run_test(submission: Submission, test: TestCase): TestResult is abstract
+	fun run_test(submission: Submission, test: TestCase): TestResult do
+		var res = new TestResult(test, submission)
+
+		var tdir = "test{submission.results.length + 1}"
+		# We get a subdirectory (a testspace) for each test case
+		#
+		# NOTE: The `as(not null)` is actually safe since `prepare_compilation` needs to be
+		# called before running tests and will fail should the workspace fail to be created,
+		# therefore returning a failure to the client.
```

Comment:
We should just assume that `prepared` means `workspace != null`.
Therefore no need to change workspace or add a method and just 

~~~nit
# bla bla
# REQUIRES `submission.workspace != null`
fun run_test(submission: Submission, test: TestCase): TestResult do
~~~